onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -398,38 +331,55 @@ common::Status SaveInitializedTensors(
-    } else if (graph.GetOrtValueInitializer(name, ort_value)) {
+      // } else if (graph.GetOrtValueInitializer(name, ort_value)) {
       // populated OrtValue from the Graph instance
     } else {
       const ONNX_NAMESPACE::TensorProto& tensor_proto = *(entry.second);
 
-      std::optional<MemBuffer> m;
+      std::optional<MemBuffer> memory_buffer;
       AllocatorPtr alloc;
       // TODO: if the tensor need be copied, does it have enough room?
-      ORT_RETURN_IF_ERROR(planner.GetPreallocatedBuffer(ort_value_index, name, m, alloc));
+      ORT_RETURN_IF_ERROR(planner.GetPreallocatedBuffer(ort_value_index, name, memory_buffer, alloc));
       bool use_device_allocator_for_initializers =
-          session_options.config_options.GetConfigOrDefault(kOrtSessionOptionsUseDeviceAllocatorForInitializers, "0") == "1";
-
-      Tensor* p_tensor = nullptr;
-      auto buffered_tensors_iter = buffered_tensors.find(name);
-      if (buffered_tensors_iter != buffered_tensors.end()) {
-        p_tensor = buffered_tensors_iter->second.get();
-      }
-
-      Status st = DeserializeTensorProto(env, graph_loc, tensor_proto, (m.has_value()) ? &*m : nullptr, alloc,
-                                         default_cpu_alloc, ort_value, data_transfer_mgr, external_data_loader_mgr,
-                                         prepacked_for_graph,
-                                         use_device_allocator_for_initializers, p_tensor);
-      if (!st.IsOK()) {
-        std::ostringstream oss;
-        oss << "Deserialize tensor " << name << " failed." << st.ErrorMessage();
-        return Status(st.Category(), st.Code(), oss.str());
-      }
-
-      if (p_tensor != nullptr) {
-        // p_tensor was wrapped in a deleter by DeserializeTensorProto so we can simply release it here.
-        ORT_IGNORE_RETURN_VALUE(buffered_tensors_iter->second.release());
-        buffered_tensors.erase(buffered_tensors_iter);
+          session_options.config_options.GetConfigOrDefault(
+              kOrtSessionOptionsUseDeviceAllocatorForInitializers, "0") == "1";
+
+      // Check if we already have an OrtValue for this initializer on CPU
+      if (OrtValue ort_value_from_graph;
+          graph.GetOrtValueInitializer(name, ort_value_from_graph)) {
+        const auto& memory_info = (alloc != nullptr) ? alloc->Info() : memory_buffer->GetAllocInfo();
+        const auto device_type = memory_info.device.Type();
+        if (device_type == OrtDevice::CPU && !memory_buffer) {
+          // This is on CPU and there is not a pre-allocated buffer for it.
+          // Use directly from the graph
+          ort_value = std::move(ort_value_from_graph);
+        } else {
+          TensorShape tensor_shape = utils::GetTensorShapeFromTensorProto(tensor_proto);
+          const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(
+                                               tensor_proto.data_type())
+                                               ->GetElementType();
```

Comment:
nit: formatting

Reply:
This is the way it formats it, cant' fight.